File: q394.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<string>st;
        string ans;
        stack<int>nums;
        int num=0;
        for(char c:s){
            if(isdigit(c)){
                num=num*10+(c-'0');
            }
            else if(isalpha(c)){
                ans.push_back(c);
            }
            else if(c=='['){
                st.push(ans);
                nums.push(num);
                ans="";
                num=0;
            }
            else if(c==']'){
                string temp=ans;
                for(int i=0;i<nums.top()-1;i++){
                    ans+=temp;
                }
                ans=st.top()+ans;
                st.pop();
                nums.pop();
            }
        }
        return ans;
    }
};

File: q394.py
class Solution:
    def decodeString(self, s: str) -> str:
        nums=[]
        st=[]
        num=0
        ans=""
        for c in s:
            if c.isdigit():
                num=num*10+int(c)  
            
            elif c.isalpha():
                ans+=c
            elif c=='[':
                st.append(ans)
                ans=""
                nums.append(num)
                num=0
            elif c==']':
                temp=ans
                ans=temp*nums[-1]
                nums.pop()
                ans=st[-1]+ans
                st.pop()
        return ans
